### ISSUE TITLE
Inconsistencies in configuration settings

### ISSUE THREAD

**Issue by paul** [bensheldon/good_job#380]
Different config settings seem to behave differently based on how they're defined. I was trying to investigate why GoodJob wasn't preserving completed jobs, even though I have the setting enabled.

In `config/initializers/good_job.rb`, I have this:

```ruby
MyApp::Application.configure do
  config.active_job.queue_adapter = :good_job

  config.good_job.tap do |good_job|
    good_job.execution_mode = :external
    good_job.preserve_job_records = true

    good_job.active_record_parent_class = "ApplicationRecord"
    good_job.on_thread_error = ->(exception) { Honeybadger.notify(exception) }

    # Enabled by the procfile on "worker", but on heroku we only want to run it
    # on the first worker dyno
    dyno = ENV["DYNO"]
    if dyno && dyno != "worker.1"
      good_job.enable_cron = false
    end

    good_job.cron = {
      refresh_slack_tokens: {
        cron:  "0 * * * *",
        class: "RefreshSlackAppTokens",
      }
    }
  end
end
```

I'm enabling `preserve_job_records`, and its set when I examine the config, but when I look at the value on GoodJob, its not set:

```
[1] pry(main)> Rails.configuration.good_job
=> {:cron=>{:refresh_slack_tokens=>{:cron=>"0 * * * *", :class=>"RefreshSlackAppTokens"}},
 :execution_mode=>:external,
 :active_record_parent_class=>"ApplicationRecord",
 :on_thread_error=>#<Proc:0x00005607b3af1160 /home/rando/Code/textus/flobot/config/initializers/good_job.rb:12 (lambda)>,
 :preserve_job_records=>true}
[2] pry(main)> GoodJob.preserve_job_records
=> false
```

Other settings set the same way, like `execution_mode`, do what I expect and change how GoodJob runs. It seems it allows `preserve_job_records` to be set on config, but it has no effect. Alternatively, if I attempt to set all these values on GoodJob directly some of them (like `GoodJob.execution_mode = :extenal`) fail with "undefined method".

It looks like GoodJob defines several `mattr_accessors`, but which settings are expected to be configured on `config`, and which on `GoodJob` directly?

**Comment by bensheldon** [comment]
Hi @paul. Thank you for opening this issue. 

The [simple answer](https://github.com/bensheldon/good_job#configuration-options) is that configuration options should be defined _before_ gem initialization in `config/application.rb` or `config/environments/{environment}.rb`. And [global options](https://github.com/bensheldon/good_job#global-options) should be put in an initializer `config/initializers/good_job.rb` that is triggered _after_ gem initialization.

I myself _try_ to organize all configuration inside of an initializer, but this is an area where I haven't figured out a good pattern to allow configuration to be defined after initialization (maybe having `GoodJob.reset_configuration!`)? I'd love your feedback and experience here.

**Comment by paul** [comment]
Hi @bensheldon, sorry its taken awhile to get back on this. I supposed I was confused about which options were "global", and which were "config", since `config.good_job` is still a "global" (there's only one of them in the Rails application). Looking at the docs, its nice that each are broken out into separate sections (which I somehow missed before), but its not obvious _why_ these different options are set in different places.

**Comment by bensheldon** [comment]
>  its not obvious why these different options are set in different places.

That's my bad as gem author. Originally there was only the configuration accessors on `GoodJob`, and then in #199 the Rails `config`-based configuration was introduced to overcome deadlocks during Rails initialization.

My thinking at the moment is that everything should go into the Rails `config`, but I'd need to build that out and overcome a few issues (like not all config working if it's updated inside of `config/initializers/*.rb` rather than `config/environments/{env}.rb`.).

**Comment by aried3r** [comment]
Hey! I'll use this issue to add my thoughts and our experiences from work as well (coming from https://github.com/bensheldon/good_job/issues/403#issuecomment-936754156).

In Rails 6.0 we added `good_job` as a dependency and used an initializer like @paul did using a `Rails.application.configure` block. This worked as expected and we were able to keep configuration in one file instead of in multiple.

Starting with Rails 6.1 the first issue we encountered was that despite having a block like this in the intializer:

```Ruby
# config/initializers/good_job.rb
Rails.application.configure do
  config.active_job.queue_adapter = :good_job
end
```

jobs were being executed using the `:async` adapter. In the console we noticed the following:

```Ruby
Rails.application.config.active_job
=> {
  :queue_adapter => :good_job
}
```

but 

```Ruby
ActiveJob::Base.queue_adapter_name
=> "async"
```

So the first option we moved to `application.rb` was `config.active_job.queue_adapter`. Then jobs that had no delay using `.perform_later` got executed with GoodJob again, however the rest of the config was not working, most notably no instances of the scheduler and notifier were being started, their `.instances` were `[]`. 
We tried a `GoodJob.restart` afterwards, in the hopes that this would perhaps reload the config, perhaps re-initialize everything (ok, I admit, I was just hoping this would magically fix it). Looking at this issue here, this is probably the not-yet-existent `GoodJob.reset_configuration` mentioned here: https://github.com/bensheldon/good_job/issues/380#issuecomment-921284574.

Anyway, I got curious and did some digging, but nothing that matched this behavior directly came up. Between Rails 6.0 and Rails 6.1 there were so many changes, nothing in particular popped out to me. However some issues on the Rails issue tracker seem like others have similar problems.

https://github.com/rails/rails/issues/36650 which links to other issues, but was closed because apparently it's a [gem issue, not a Rails issue](https://github.com/rails/rails/issues/36650#issuecomment-511632494) (which I don't think applies to GoodJob, since it ties in deeply into Rails configuration). However it was later clarified in a code comment in newly generated Rails apps that initializers are just too late to configure the Rails app, see https://github.com/rails/rails/pull/34985.

There's also https://github.com/rails/rails/issues/37270 which I think is not directly related since according to some people it's still unresolved but it got me looking in the right direction.

Also found https://github.com/rails/rails/issues/40552 which _feels_ like it should be related, but I couldn't tell.

TL;DR: 
> "In general, configuration specific to the application object needs to go either in config/application.rb or in config/environments/#{Rails.env}.rb, you can't just assign to an arbitrary config point in config/initializers and expect it to work." <sup>https://github.com/rails/rails/issues/24748#issuecomment-216317145</sup>

So, we did what GoodJob's README says and put everything into `config/application.rb` and everything now works as expected. We just left the happy path and perhaps were using undefined behavior in Rails 6.0.

**Comment by bensheldon** [comment]
I think this problem has been addressed, mainly with #454 and #460. Configuration should now primarily be set via config and it doesn't matter if it's done in an initializer or within {application,environment}.rb files
